lib/tokens.ts

```diff
@@ -0,0 +1,101 @@
+import { PublicKey } from "@solana/web3.js";
+import { TokenProps } from "./types";
+import {
+  JsonMetadata,
+  MPL_TOKEN_METADATA_PROGRAM_ID,
+  deserializeMetadata,
+} from "@metaplex-foundation/mpl-token-metadata";
+import { RpcAccount } from "@metaplex-foundation/umi";
+/**
+ * Starts with the jup.ag strict list to find token. jup.ag maintains a list of quality tokens
+ * if that fails, use metaplex with RPC call and fetch metadata json
+ * if THAT fails, take MINT address and truncate by 5 characters and that's the symbol (no image)
+ */
+export async function enrichTokenMetadata(
+  tokenAddress: PublicKey
+): Promise<TokenProps> {
+  // first check strict list
+  const tokenOnJup = await getTokenFromJupStrictList(tokenAddress);
+  if (tokenOnJup) {
+    return {
+      symbol: tokenOnJup.symbol,
+      publicKey: tokenAddress.toString(),
+      url: tokenOnJup.logoURI,
+    };
+  }
+
+  // try metaplex
+  const jsonMetadata = await getMetaplexMetadataForToken(tokenAddress);
+  if (jsonMetadata && jsonMetadata.symbol) {
+    return {
+      symbol: jsonMetadata.symbol,
+      publicKey: tokenAddress.toString(),
+      url: jsonMetadata.image,
+    };
+  }
+
+  // finally just return truncated address for symbol
+  return {
+    symbol: tokenAddress.toString().slice(0, 5).toUpperCase(),
+    publicKey: tokenAddress.toString(),
+  };
+}
+
+type Token = {
+  address: string;
+  chainId: number;
+  decimals: number;
+  name: string;
+  symbol: string;
+  logoURI: string;
+  tags: string[];
+  extensions: {
+    coingeckoId: string;
+  };
+};
+
+async function getTokenFromJupStrictList(
+  address: PublicKey
+): Promise<Token | null> {
+  try {
+    const response = await fetch("https://token.jup.ag/strict");
```

Comment:
We can pull this locally and sync, or fetch from GitHub if we need to. Either way it fails over to download it locally (IMO) or we just keep DB sync'd. Fetching this multiple times would not be ideal.

Reply:
Okay, I implemented a github fallback for this.